controllers/topology/internal/extensions/patches/inline/json_patch_generator.go

```diff
@@ -0,0 +1,293 @@
+	"strings"
+	"text/template"
+
+	"github.com/pkg/errors"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/extensions/patches/api"
+	"sigs.k8s.io/yaml"
+)
+
+// JSONPatchGenerator generates JSON patches for a GenerateRequest based on a ClusterClassPatch.
+type JSONPatchGenerator struct {
+	patch *clusterv1.ClusterClassPatch
+}
+
+// NewJSONPatchGenerator returns a JSONPatchGenerator.
+func NewJSONPatchGenerator(patch *clusterv1.ClusterClassPatch) api.Generator {
+	return &JSONPatchGenerator{
+		patch: patch,
+	}
+}
+
+// Generate generates JSON patches for the given GenerateRequest based on a ClusterClassPatch.
+func (j *JSONPatchGenerator) Generate(_ context.Context, req *api.GenerateRequest) (*api.GenerateResponse, error) {
+	resp := &api.GenerateResponse{}
+
+	// Loop over all templates.
+	for _, template := range req.Items {
+		// Loop over all PatchDefinitions.
+		for _, patch := range j.patch.Definitions {
+			// Return if the template does not match the selector.
+			if !templateMatchesSelector(&template.TemplateRef, patch.Selector) {
+				continue
+			}
+
+			// Merge template-specific and global variables.
+			variables := mergeVariables(template.Variables, req.Variables)
+
+			// Generate JSON patches.
+			jsonPatches, err := generateJSONPatches(patch.JSONPatches, variables)
+			if err != nil {
+				return nil, err
+			}
+
+			// Add jsonPatches to the response.
+			resp.Items = append(resp.Items, api.GenerateResponsePatch{
+				TemplateRef: template.TemplateRef,
+				Patch:       *jsonPatches,
+				PatchType:   api.JSONPatchType,
+			})
+		}
+	}
+
+	return resp, nil
+}
+
+// templateMatchesSelector returns true if the template matches the selector.
+func templateMatchesSelector(templateRef *api.TemplateRef, selector clusterv1.PatchSelector) bool {
+	// Check if the apiVersion and kind are matching.
+	if templateRef.APIVersion != selector.APIVersion {
+		return false
+	}
+	if templateRef.Kind != selector.Kind {
+		return false
+	}
+
+	// Check if target matches.
+	switch templateRef.TemplateType {
+	case api.InfrastructureClusterTemplateType:
+		// Check if matchSelector.infrastructureCluster is true.
+		if selector.MatchResources.InfrastructureCluster == nil {
+			return false
+		}
+		return *selector.MatchResources.InfrastructureCluster
+	case api.ControlPlaneTemplateType, api.ControlPlaneInfrastructureMachineTemplateType:
+		// Check if matchSelector.controlPlane is true.
+		if selector.MatchResources.ControlPlane == nil {
+			return false
+		}
+		return *selector.MatchResources.ControlPlane
+	case api.MachineDeploymentBootstrapConfigTemplateType, api.MachineDeploymentInfrastructureMachineTemplateType:
+		// Check if matchSelector.machineDeploymentClass.names contains the
+		// MachineDeployment.Class of the template.
+		if selector.MatchResources.MachineDeploymentClass == nil {
+			return false
```

Comment:
Q: if we changes this should we get support for patches applying to all machine classes?

Reply:
I think we would add the following in line 110:
```go
		if len(selector.MatchResources.MachineDeploymentClass.Names) == 0 {
			return true
		}
```

This would trigger with the following selector
```yaml
        matchResources:
          machineDeploymentClass:
            names: []
```

Reply:
Expressing an empty array for getting all is kind of weird.
We can add support for `names: ['*']` but if this is slowing down this PR let's open an issue and track this as improvement

Reply:
Let's move this to a follow up PR given that this is on the finish line

Reply:
It would definitely be weird. I think your suggestion has a nice similarity with how it's done in Roles/ClusterRoles so +1.

Reply:
@fabriziopandini created: https://github.com/kubernetes-sigs/cluster-api/issues/5648

Might qualify for help wanted, not sure. Your decision :)